src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionMetadata.java

```diff
@@ -311,73 +299,52 @@ private static Optional<RootModuleFileFixup> generateFixup(
 
     message += "Fix the use_repo calls by running 'bazel mod tidy'.";
 
-    var buildozerCommands =
-        Stream.of(
-                makeUseRepoCommand(
-                    "use_repo_add",
-                    false,
-                    importsToAdd,
-                    extensionBzlFile,
-                    extensionName,
-                    rootUsage.getIsolationKey()),
-                makeUseRepoCommand(
-                    "use_repo_remove",
-                    false,
-                    importsToRemove,
-                    extensionBzlFile,
-                    extensionName,
-                    rootUsage.getIsolationKey()),
-                makeUseRepoCommand(
-                    "use_repo_add",
-                    true,
-                    devImportsToAdd,
-                    extensionBzlFile,
-                    extensionName,
-                    rootUsage.getIsolationKey()),
-                makeUseRepoCommand(
-                    "use_repo_remove",
-                    true,
-                    devImportsToRemove,
-                    extensionBzlFile,
-                    extensionName,
-                    rootUsage.getIsolationKey()))
-            .flatMap(Optional::stream)
-            .collect(toImmutableList());
+    var moduleFilePathToCommandsBuilder = ImmutableListMultimap.<PathFragment, String>builder();
+    // Repos to add are easy: always add them to the first proxy of the correct type.
+    if (!importsToAdd.isEmpty()) {
+      Proxy firstNonDevProxy =
+          rootUsage.getProxies().stream().filter(p -> !p.isDevDependency()).findFirst().get();
+      moduleFilePathToCommandsBuilder.put(
+          firstNonDevProxy.getContainingModuleFilePath(),
+          makeUseRepoCommand("use_repo_add", firstNonDevProxy.getProxyName(), importsToAdd));
+    }
+    if (!devImportsToAdd.isEmpty()) {
+      Proxy firstDevProxy =
+          rootUsage.getProxies().stream().filter(p -> p.isDevDependency()).findFirst().get();
+      moduleFilePathToCommandsBuilder.put(
+          firstDevProxy.getContainingModuleFilePath(),
+          makeUseRepoCommand("use_repo_add", firstDevProxy.getProxyName(), devImportsToAdd));
+    }
+    // Repos to remove are a bit trickier: remove them from the proxy that actually imported them.
```

Comment:
Since we are not failing on import errors, this could actually be multiple proxies for one repo.

Reply:
are we not? pretty sure we just throw a regular EvalException when there's repo name collision (whether or not they're from use_repo or for example bazel_dep). Do we filter them out later on?

Reply:
You are right!